toolkit/tools/internal/timestamp_v2/timestamp_v2_mgr.go

```diff
@@ -0,0 +1,620 @@
+		}
+		mgr.lastUpdate = time.Now()
+		mgr.shouldWrite = false
+	}
+}
+
+// BeginTiming will begin collecting fresh timing data for a high level component 'toolName' into the file at `outputFile`.
+// This function synchronizes with ReadTimingData(), ResumeTiming(), EndTiming() to ensure there is no data partially
+// written to the file. Expected weight is used to inform the dashboarding tools how many substeps the root node is expecting to have.
+// The function will NEVER return a nil TimeStamp pointer, in event of an error it will return an empty timestamp object instead.
+func BeginTiming(toolName, outputFile string, expectedWeight float64, atomicOperation bool) (node *TimeStamp, err error) {
+	if err = ensureNoManager(); err != nil {
+		return StampMgr.activeNode, err
+	}
+
+	// Create a file to write data to
+	outFileDescriptor, err := os.OpenFile(outputFile, os.O_CREATE|os.O_RDWR, 0664)
+	if err != nil {
+		err = fmt.Errorf("unable to create file %s: %w", outputFile, err)
+		failsafeLoggerWarnf(err.Error())
+		return &TimeStamp{}, err
+	}
+	// We will not close the file handle here unless there is an error, we will use it for synchronization going forwards.
+
+	// We will keep a shared lock on this file until we are done with timing. Grab in exclusive mode for now while we initialize the file
+	if err = ensureExclusiveFileLockOnFile(outFileDescriptor); err != nil {
+		outFileDescriptor.Close()
+		return &TimeStamp{}, err
+	}
+
+	// Create a new instance of the StampMgr global and initialize it with an empty root node.
+	root, err := createTimeStamp(toolName, time.Now(), expectedWeight)
+	if err != nil {
+		err = fmt.Errorf("unable to initialize timing root %s: %w", toolName, err)
+		failsafeLoggerWarnf(err.Error())
+		// Need to release the lock and file before we fail out
+		unlockFileLock(outFileDescriptor)
+		outFileDescriptor.Close()
+		return &TimeStamp{}, err
+	}
+
+	StampMgr = newStampMgr(root, outputFile, outFileDescriptor, atomicOperation)
+	// We already have the flock, make sure to grab the mutex as well so we can gracefully unlock it later
+	StampMgr.lock.Lock()
+	StampMgr.updateFile()
+	if err = StampMgr.releaseWriteLocks(); err != nil {
+		return &TimeStamp{}, fmt.Errorf("could not begin timing: %w", err)
+	}
+	return StampMgr.activeNode, err
+}
+
+// ResumeTiming will begin appending timing data for a high level component 'toolName' into the file at `outputFile`.
+// This function synchronizes with ReadTimingData(), BeginTiming(), EndTiming()  to ensure there is no data partially
+// written to the file. The function will NEVER return a nil TimeStamp pointer, in event of an error it will return an empty
+// timestamp object instead. This function will NOT update the currently active node, it will remain pointing to the root node.
+func ResumeTiming(toolName, existingTimingFile string, createIfMissing bool, atomicOperation bool) (node *TimeStamp, err error) {
```

Comment:
We need a way to resume from disk since there are some tasks we do want to get timing data for (toolchian build). The way I managed this was to write a small go program that reads, updates the timestamps, then writes back the json file. This can be invoked from the shell scripts.